### PR TITLE
[CUDA] 64-bit indexing fixes for cross-entropy kernels

### PR DESCRIPTION
**Pull request by eqy**
For #108345, #111484

Addresses the forward kernels implicated in the issues, but will take another look at the backward kernels (in follow-up PRs if necessary).

The spatial softmax kernel is changed to use signed integer indexing rather than unsigned as `ScalarType` only has signed integer types declared for now, but this should be a minor change.

CC @ptrblck @crcrpar (who landed a few related PRs recently).

cc @ptrblck

### CODE REVIEW

**Review by mikaylagawarecki — APPROVED** [review]
Thanks, please fix lint!

**Comment by mikaylagawarecki** [comment]
Hey, looks like the lint is still failing :/
